src/mame/brother/lw350.cpp

```diff
@@ -0,0 +1,1749 @@
+	#ifndef UPD765
+		LW350_FLOPPY_CONNECTOR(config, floppy, 0);
+		floppy->option_add("35hd", LW350_FLOPPY);
+		floppy->set_default_option("35hd");
+	#else
+		FLOPPY_CONNECTOR(config, floppy, lw350_floppies, "35hd", floppy_image_device::default_pc_floppy_formats);
+	#endif
+
+	HD63266F(config, fdc, XTAL(16'000'000));
+	fdc->irq_wr_cb().set(FUNC(lw350_state::fdc_irq_w));
+	fdc->dreq_wr_cb().set(FUNC(lw350_state::fdc_dreq_w));
+	#ifndef UPD765
+		fdc->dend_rd_cb().set(FUNC(lw350_state::fdc_dend_r));
+	#endif
+
+	// sound hardware
+	SPEAKER(config, "mono").front_center();
+	BEEP(config, "beeper", 4'000).add_route(ALL_OUTPUTS, "mono", 1.0); // 4.0 kHz
+}
+
+//////////////////////////////////////////////////////////////////////////
+// LW-450
+//////////////////////////////////////////////////////////////////////////
+
+/***************************************************************************
+
+Brother LW-450
+1992
+
+Hardware:
+
+#13
+Hitachi
+HD64180RF6X
+8-bit CMOS Micro Processing Unit
+fully compatible with Zilog Z80180 (Z180)
+6 MHz, FP-80B, Address Space 1 M Byte
+MuRata CST12MTW 12.00 MHz Ceramic Resonator
+
+#12
+Mitsubishi
+M65020FP
+UA7777-A
+
+#15
+Hitachi
+HD74LS368AP
+
+#10, #11
+Hitachi
+HD74LS157P
+
+#7, #8
+NEC
+D41464C-10
+65,536 x 4-Bit
+Dynamic NMOS RAM
+100 ns
+18-pin plastic DIP
+
+#1
+NEC
+D23C4001EC-172
+UA2849-A
+4MBit Mask ROM
+
+#2
+AMD
+AM27C020
+2BC04
+2MBit (256K x 8-Bit) CMOS EPROM
+
+#3
+Hitachi
+HM65256BSP-10
+32,768 word x 8-bit High Speed Pseudo Static RAM
+100 ns
+DP-28N
+
+#4
+Hitachi HD63266F
+CMOS Floppy Disk Controller
+QFP-64
+Murata Ceralock CST-MXW 16.00 MHz Ceramic Resonator
+
+#5
+Hitachi
+HD6445P4
+CRTC-II (CRT Controller)
+DP-40
+80 system Bus Interface
+4.0 MHz Bus Timing
+
+#9
+Mitsubishi
+M65133FP
+UA7550-A
+
+#14
+Hitachi
+HM62256AF-8
+32,768 word x 8-bit High Speed CMOS Static RAM
+85 ns
+FP-28
+
+720kb Floppy Drive
+
+D-SUB9 CRT Connector
+HSYNC 18.5 kHz
+VSYNC 50 Hz
+PC MDA standard
+  DB9 connector 1+2 GND, 3+4+5 nc, 6 Intensity, 7 Brightness, 8 HSync, 9 VSync
+  HSync positive, VSync negative active
+
+Emulation Status:
+Printer not working
+Floppy Read has some problems (directory working, but LW-450 reports illegal file format when trying to load a .wpt file (content verified with LW-350), but writing seems fine)
+Dictionary ROM probably not correctly mapped
+
+***************************************************************************/
+
+constexpr int MDA_CLOCK = 16'257'000;
+
+class lw450_state : public driver_device
+{
+public:
+	lw450_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		maincpu(*this, "maincpu"),
+		screen(*this, "screen"),
+		palette(*this, "palette"),
+		floppy(*this, "floppy"),
+		fdc(*this, "fdc"),
+		m_crtc(*this, "hd6445"),
+		vram(*this, "vram"),
+		speaker(*this, "beeper"),
+		io_kbrow(*this, "kbrow.%u", 0),
+		rom(*this, "maincpu")
+	{ }
```

Comment:
Can the two driver state classes use a common base class, or do they really have nothing in common?

Reply:
currently they only share keyboard and floppy, different video, different memory map, different roms.

Reply:
although the keyboard should probably be shared between LW-30 and LW-450 rather.